### PR TITLE
Implementation of KubeEdge Command line tool "kectl".

### LINKED ISSUE #360: Can we add a node by a single command?
**What would you like to be added**:
Can we add a node by a single command like “k3s agent -s ${SERVER_URL} -t ${NODE_TOKEN} ” in k3s, or “docker swarm join-token ${NODE_TOKEN}” in docker swarm?
**Why is this needed**:
The operation is simple and common.

### PR DESCRIPTION
**Pull request by samy2019**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Many users complained that kubeEdge installation is too complicated and it may prevent people from trying kubeEdge. There should a simplified way to have Getting Started with KubeEdge, so that user can concentrate more on using it instantly, rather than getting entangled in the installation steps.
KubeEdge requires a series of pre-installations, which at times are challenging for few user.
To give a better experience and usability, "kectl" is implemented. 

**Which issue(s) this PR fixes**:
Fixes #342 
Fixes #360 

**Special notes for your reviewer**:
1. This PR is raised for edge node commands like, "kectl edge join" and "kectl edge reset" on ubuntu OS.
2. README.md has to be updated, which will be done at later point of time.
3. Code comments are yet to be added every where.
4. Cloud commands like "kectl cloud init" and "kectl cloud reset" will be added to this PR soon.
5. There will be lot of debug fmt.Println's, please ignore them as of now. It will be removed soon.
6. Lints will fail, so please ignore for now.

### CODE REVIEW

**Comment by sids-b** [comment]
please let me know when this PR is ready for the 1st round of review.

**Comment by sids-b** [comment]
@samy2019 , please don't update the existing vendor packages. If they are auto updating when you run dep ensure, please add a override in gopkg.toml. Something like this
``` 
[[override]]
  name = "github.com/go-chassis/paas-lager"
  revision = "eff93e5e67dbf7eab20c058db4e37e8739bf8df5"
```

**Comment by samy2019** [comment]
**Special notes for your reviewer**:
1. This PR is raised for edge node commands like, "kectl edge join" and "kectl edge reset" on ubuntu OS.
2. README.md has to be updated, which will be done at later point of time. - This is in progress.
3. Code comments are yet to be added every where. Completed
4. Cloud commands like "kectl cloud init" and "kectl cloud reset" will be added to this PR soon. - PR #435 raised
5. There will be lot of debug fmt.Println's, please ignore them as of now. It will be removed soon. - Removed
6. Lints will fail, so please ignore for now. - Passing now, It was due to point 3 above

**Comment by m1093782566** [comment]
I would appreciate you squash the commits and seperate the commits again - one for vendor and one for main changes.

File: kectl/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# KubeEdge Installer
+
+[Kubeedge Installer Doc](<https:https://github.com/samy2019/kubeedge/blob/bc14b1ee9f71b023b1e9a6faf68a0ee01ff35450/docs/proposals/kubeedgeadm-scope.md/>)
```
Comment (rohitsardesai83) [inline kectl/README.md]: please correct this link . it should not point to personal fork.

File: kectl/app/cmd/cloud.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/*
```
Comment (sids-b) [inline kectl/app/cmd/cloud.go]: These pages looks like a duplicate from pr #435. Please delete from 1 place.
The PR description says that this PR is only for edge commands so I guess these pages should be deleted from this PR
cc @srivatsav123

File: kectl/app/cmd/edge/join.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  - This command will download and install the default version of pre-requisites and KubeEdge
+
+kectl edge join --server=10.20.30.40:8080 --docker-version= --kubeedge-version=0.2.1 --kubernetes-version=1.14.1
+kectl edge join --server=10.20.30.40:8080 --docker-version --kubeedge-version=0.2.1 --kubernetes-version=1.14.1
+
+  - In case, any option is used in a format like as shown for "--docker-version" or "--docker-version=", without a value
+		then default values will be used. 
+		Also options like "--docker-version", "--kubernetes-version" and "--kubeedge-version", version should be in 
+		format like "18.06.3", "1.14.0" and "0.2.1".
+`
```
Comment (sids-b) [inline kectl/app/cmd/edge/join.go]: please remove blank line

File: kectl/app/cmd/edge/join.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+further instructions and forward telemetry data from 
+devices to cloud
+`
+	edgeJoinExample = `
+kectl edge join --server=<ip:port>
+
+  - For this command --server option is a Mandatory option
+  - This command will download and install the default version of pre-requisites and KubeEdge
+
+kectl edge join --server=10.20.30.40:8080 --docker-version= --kubeedge-version=0.2.1 --kubernetes-version=1.14.1
```
Comment (sids-b) [inline kectl/app/cmd/edge/join.go]: Why do we need to specify kubernetes version at edge ??

File: kectl/app/cmd/util/edgecoreinstaller.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+//TearDown method will remove the edge node from api-server and stop edge_core process
+func (ku *KubeEdgeInstTool) TearDown() error {
+
+	ku.SetOSInterface(GetOSInterface())
+
+	//Remove the edge from api server using kubectl command, like below
+	//kubectl delete -f $GOPATH/src/github.com/kubeedge/kubeedge/build/node.json -s http://192.168.20.50:8080
+	nodeJSONApply := fmt.Sprintf("kubectl delete -f %s -s http://%s", KubeEdgeConfigNodeJSON, ku.K8SApiServerIP)
```
Comment (sids-b) [inline kectl/app/cmd/util/edgecoreinstaller.go]: nodeJSONApply -> nodeJSONDelete ?

File: kectl/app/cmd/util/edgecoreinstaller.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+		return fmt.Errorf("%s", errout)
+	}
+	fmt.Println(cmd.GetStdOutput())
+
+	return nil
+}
+
+//TearDown method will remove the edge node from api-server and stop edge_core process
+func (ku *KubeEdgeInstTool) TearDown() error {
+
```
Comment (sids-b) [inline kectl/app/cmd/util/edgecoreinstaller.go]: please remove blank lines at the beginning and end of functions
Applicable to all similar places

File: kectl/app/cmd/util/k8sinstaller.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+//K8SInstTool embedes Common struct and contains the default K8S version and
+//a flag depicting if host is an edge or cloud node
+//It implements ToolsInstaller interface
+type K8SInstTool struct {
+	Common
+	IsEdgeNode     bool //True - Edgenode False - Cloudnode
+	DefaultToolVer string
+}
+
+//InstallTools sets the OS interface, checks if K8S installation is required or not.
```
Comment (sids-b) [inline kectl/app/cmd/util/k8sinstaller.go]: is k8sInstallation required on edge node ? Even if is required, it should be handled by cloud command I guess ?

File: kectl/app/cmd/util/ubuntuinstaller.go [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+
+	return VersionNAInRepo, nil
+}
+
+//InstallK8S will install kubeadm, kudectl and kubelet for the cloud node
+//and only kubectl for edge node
+func (u *UbuntuOS) InstallK8S() error {
+	kcomp := fmt.Sprintf("Installing %s version of ", u.KubernetesVersion)
+	if u.IsEdgeNode == true {
+		kcomp = kcomp + "kubectl"
```
Comment (sids-b) [inline kectl/app/cmd/util/ubuntuinstaller.go]: I guess kubectl installation is not required on edgeNode

File: kectl/app/cmd/edge/join.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+				fmt.Println(err.Error())
+				continue
+			}
+		}
+	}
+
+	//Install and Start KubeEdge Node
+	err := toolList["KubeEdge"].InstallTools()
+	if err != nil {
+		fmt.Println(err.Error())
```
Comment (sids-b) [inline kectl/app/cmd/edge/join.go]: can we use logs instead of fmt.Print wherever applicable ?

**Comment by sids-b** [comment]
As per discussion with @srivatsav123 and @samy2019 , new PR will be raised

**Pull request closed without merging**